packages/appium/docs/en/ecosystem/build-drivers.md

```diff
@@ -2,4 +2,617 @@
+call something like:
+
+```js
+// webdriverio example
+driver.executeScript('soundz: playSong', [{song: 'Stairway to Heaven', artist: 'Led Zeppelin'}]);
+```
+
+Then in your driver code you could simply look for `soundz:` at the beginning of your
+`execute` override, and do the appropriate thing:
+
+```js
+async execute(script, args) {
+  if (script === 'soundz: playSong') {
+    const {song, artist} = args[0];
+    // play the song based on song and artist details
+    return;
+  }
+  // otherwise do the normal thing, or throw a not yet implemented
+}
+```
+
+### Implement handling of Appium settings
+
+Appium users can send parameters to your driver via CLI args as well as via capabilities. But these
+cannot change during the course of a test, and sometimes users want to adjust parameters mid-test.
+Appium has a [Settings](../guides/settings.md) API for this purpose.
+
+To support settings in your own driver, first of all define `this.settings` to be an instance of
+the appropriate class, in your constructor:
+
+```js
+// import {DeviceSettings} from 'appium/driver';
+
+this.settings = new DeviceSettings();
+```
+
+Now, you can read user settings any time simply by calling `this.settings.getSettings()`. This will
+return a JS object where the settings names are keys and have their corresponding values.
+
+If you want to assign some default settings, or run some code on your end whenever settings are
+updated, you can do both of these things as well.
+
+```js
+constructor() {
+  const defaults = {setting1: 'value1'};
+  this.settings = new DeviceSettings(defaults, this.onSettingsUpdate.bind(this));
+}
+
+async onSettingsUpdate(key, value) {
+  // do anything you want here with key and value
+}
+```
+
+### Make itself aware of resources other concurrent drivers are using
+
+Let's say your driver uses up some system resources, like ports. There are a few ways to make sure
+that multiple simultaneous sessions don't use the same resources:
+
+1. Have your users specify resource IDs via capabilities (`appium:driverPort` etc)
+1. Just always use free resources (find a new random port for each session)
+1. Have each driver express what resources it is using, then examine currently-used resources from
+   other drivers when a new session begins.
+
+To support this third strategy, you can implement `get driverData` in your driver to return what
+sorts of resources your driver is currently using, for example:
+
+```
+get driverData() {
+  return {specialPort: 1234, specialFile: /path/to/file}
+}
+```
+
+Now, when a new session is started on your driver, the `driverData` response from any other
+simultaneously running drivers will also be included, as the last parameter of the `createSession`
+method:
+
+```
+async createSession(jwpCaps, reqCaps, w3cCaps, driverData)
+```
+
+You can dig into this `driverData` array to see what resources other drivers are using to help
+determine which ones you want to use for this particular session.
```

Comment:
I think I was a little confused about this, and I'm not sure what would be an accurate suggestion, but this seems to mean "instances of other drivers", and naturally those are per-session, and may an instance of your driver or some other driver

Reply:
actually it's just drivers of this driver's type. i'll add a note.